File: app/src/main/java/com/stathis/elmepaunivapp/ui/chatbot/recyclerview/ChatBotAdapter.java
package com.stathis.elmepaunivapp.ui.chatbot.recyclerview;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.stathis.elmepaunivapp.R;
import com.stathis.elmepaunivapp.abstraction.DiffItemCallbackClass;
import com.stathis.elmepaunivapp.listeners.ItemClickListener;
import com.stathis.elmepaunivapp.ui.chatbot.model.Answer;
import com.stathis.elmepaunivapp.ui.chatbot.model.ChatbotHeader;
import com.stathis.elmepaunivapp.ui.chatbot.model.Question;

public class ChatBotAdapter extends ListAdapter<Object, RecyclerView.ViewHolder> {

    private ItemClickListener itemClickListener;

    public ChatBotAdapter(@NonNull ItemClickListener itemClickListener) {
        super(new DiffItemCallbackClass<Object>());
        this.itemClickListener = itemClickListener;
    }


    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(viewType, parent, false);
        if (viewType == R.layout.holder_question_item) {
            return new MessageViewHolder(view);
        } else if (viewType == R.layout.holder_answer_item){
            return new AnswerViewHolder(view, itemClickListener);
        } else if (viewType == R.layout.holder_chatbot_header){
            return new ChatbotHeaderViewHolder(view);
        } else {
            return null;
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (getItem(position) instanceof Question) {
            ((MessageViewHolder) holder).present((Question) getItem(position));
        } else if (getItem(position) instanceof Answer) {
            ((AnswerViewHolder) holder).present((Answer) getItem(position));
        } else if (getItem(position) instanceof ChatbotHeader) {
            ((ChatbotHeaderViewHolder) holder).present((ChatbotHeader) getItem(position));
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (getItem(position) instanceof Question) {
            return R.layout.holder_question_item;
        } else if (getItem(position) instanceof Answer) {
            return R.layout.holder_answer_item;
        } else if (getItem(position) instanceof ChatbotHeader) {
            return R.layout.holder_chatbot_header;
        } else {
            return 3;
        }
    }
}

File: app/src/main/java/com/stathis/elmepaunivapp/ui/chatbot/recyclerview/ChatbotAdapter.kt
package com.stathis.elmepaunivapp.ui.chatbot.recyclerview

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.ListAdapter
import com.stathis.elmepaunivapp.R
import com.stathis.elmepaunivapp.abstraction.DiffItemClass
import com.stathis.elmepaunivapp.callbacks.ElmepaClickListener
import com.stathis.elmepaunivapp.abstraction.LocalModel
import com.stathis.elmepaunivapp.databinding.HolderAnswerItemBinding
import com.stathis.elmepaunivapp.databinding.HolderChatbotHeaderBinding
import com.stathis.elmepaunivapp.databinding.HolderEmptyItemRowBinding
import com.stathis.elmepaunivapp.databinding.HolderQuestionItemBinding
import com.stathis.elmepaunivapp.ui.chatbot.model.Answer
import com.stathis.elmepaunivapp.ui.chatbot.model.ChatbotHeader
import com.stathis.elmepaunivapp.ui.chatbot.model.Question

class ChatbotAdapter(private val callback : ElmepaClickListener) : ListAdapter<LocalModel, ChatbotViewHolder>(DiffItemClass<LocalModel>()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ChatbotViewHolder = when(viewType){
        R.layout.holder_question_item -> {
            val view = HolderQuestionItemBinding.inflate(LayoutInflater.from(parent.context),parent,false)
            ChatbotViewHolder(view,callback)
        }
        R.layout.holder_answer_item -> {
            val view = HolderAnswerItemBinding.inflate(LayoutInflater.from(parent.context),parent,false)
            ChatbotViewHolder(view,callback)
        }
        R.layout.holder_chatbot_header -> {
            val view = HolderChatbotHeaderBinding.inflate(LayoutInflater.from(parent.context),parent,false)
            ChatbotViewHolder(view,callback)
        }
        else -> {
            val view = HolderEmptyItemRowBinding.inflate(LayoutInflater.from(parent.context),parent,false)
            ChatbotViewHolder(view,callback)
        }
    }

    override fun onBindViewHolder(holder: ChatbotViewHolder, position: Int) = holder.bindData(getItem(position))

    override fun getItemViewType(position: Int): Int = when(getItem(position)){
        is Question -> R.layout.holder_question_item
        is Answer -> R.layout.holder_answer_item
        is ChatbotHeader -> R.layout.holder_chatbot_header
        else -> R.layout.holder_empty_item_row
    }
}